api/helpers.go

```diff
@@ -365,38 +407,13 @@ func loadCAPool(config *Config) (*x509.CertPool, error) {
 
 // Do makes an HTTP(S) request with predefined http.Request object.
 // Caller is responsible for calling http.Response.Body().Close()
-func Do(req *http.Request, timeout time.Duration, caPool *x509.CertPool, headers map[string]string) (resp *http.Response, err error) {
-	client := http.Client{
-		Timeout: timeout,
-	}
-
-	// go http client does not copy the headers when it follows the redirect.
-	// https://github.com/golang/go/issues/4800
-	client.CheckRedirect = func(req *http.Request, via []*http.Request) error {
-		for attr, val := range via[0].Header {
-			if _, ok := req.Header[attr]; !ok {
-				req.Header[attr] = val
-			}
-		}
-		return nil
-	}
+func Do(req *http.Request, timeout time.Duration, headers map[string]string, client *http.Client, secureTransport *http.Transport) (resp *http.Response, err error) {
+	client.Timeout = timeout
 
 	if req.URL.Scheme == "https" {
-		var tlsClientConfig *tls.Config
-		if caPool == nil {
-			// do HTTPS without certificate verification.
-			tlsClientConfig = &tls.Config{
-				InsecureSkipVerify: true,
-			}
-		} else {
-			tlsClientConfig = &tls.Config{
-				RootCAs: caPool,
-			}
-		}
-
-		client.Transport = &http.Transport{
-			TLSClientConfig: tlsClientConfig,
-		}
+		client.Transport = secureTransport
```

Comment:
don't think you should be swapping out Transports inside a client object. this feels very racy.